community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java

```diff
@@ -0,0 +1,354 @@
+    }
+
+    private static boolean compareAndSetState( long address, long expect, long update )
+    {
+        return UnsafeUtil.compareAndSwapLong( null, address, expect, update );
+    }
+
+    private static void unconditionallySetState( long address, long update )
+    {
+        UnsafeUtil.putLongVolatile( address, update );
+    }
+
+    /**
+     * Start an optimistic critical section, and return a stamp that can be used to validate if the read lock was
+     * consistent. That is, if no write or exclusive lock was overlapping with the optimistic read lock.
+     *
+     * @return A stamp that must be passed to {@link #validateReadLock(long,long)} to validate the critical section.
+     */
+    public static long tryOptimisticReadLock( long address )
+    {
+        return getState( address ) & SEQ_MASK;
+    }
+
+    /**
+     * Validate a stamp from {@link #tryOptimisticReadLock(long)} or {@link #unlockExclusive(long)}, and return {@code true}
+     * if no write or exclusive lock overlapped with the critical section of the optimistic read lock represented by
+     * the stamp.
+     *
+     * @param stamp The stamp of the optimistic read lock.
+     * @return {@code true} if the optimistic read lock was valid, {@code false} otherwise.
+     */
+    public static boolean validateReadLock( long address, long stamp )
+    {
+        UnsafeUtil.loadFence();
+        return (getState( address ) & CHK_MASK) == stamp;
+    }
+
+    public static boolean isModified( long address )
+    {
+        return (getState( address ) & MOD_MASK) == MOD_MASK;
+    }
+
+    public static boolean isExclusivelyLocked( long address )
+    {
+        return (getState( address ) & EXL_MASK) == EXL_MASK;
+    }
+
+    /**
+     * Try taking a concurrent write lock. Multiple write locks can be held at the same time. Write locks will
+     * invalidate any optimistic read lock that overlaps with them, and write locks will make any attempt at grabbing
+     * an exclusive lock fail. If an exclusive lock is currently held, then the attempt to take a write lock will fail.
+     * <p>
+     * Write locks must be paired with a corresponding {@link #unlockWrite(long)}.
+     *
+     * @return {@code true} if the write lock was taken, {@code false} otherwise.
+     */
+    public static boolean tryWriteLock( long address )
+    {
+        long s, n;
+        for (; ; )
+        {
+            s = getState( address );
+            boolean unwritablyLocked = (s & EXL_MASK) != 0;
+            boolean writeCountOverflow = (s & CNT_MASK) == CNT_MASK;
+
+            // bitwise-OR to reduce branching and allow more ILP
+            if ( unwritablyLocked | writeCountOverflow )
+            {
+                return failWriteLock( s, writeCountOverflow );
+            }
+
+            n = s + CNT_UNIT | MOD_MASK;
+            if ( compareAndSetState( address, s, n ) )
+            {
+                UnsafeUtil.storeFence();
```

Comment:
No, this is to prevent plain stores within the critical section from reordering _into_ the compareAndSetState, which is theoretically possible, e.g. if it is implemented with LL/SC, or with a mutex. The plain store could then reorder with the store that is the taking of the lock, thus effectively escaping the critical section.